modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/upgrade/v3_7_1/test/DDMStructureEmptyValidationUpgradeProcessTest.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.dynamic.data.mapping.upgrade.v3_7_1.test;
+
+import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
+import com.liferay.change.tracking.test.util.BaseCTUpgradeProcessTestCase;
+import com.liferay.dynamic.data.lists.model.DDLRecordSet;
+import com.liferay.dynamic.data.mapping.form.field.type.constants.DDMFormFieldTypeConstants;
+import com.liferay.dynamic.data.mapping.io.DDMFormSerializer;
+import com.liferay.dynamic.data.mapping.model.DDMForm;
+import com.liferay.dynamic.data.mapping.model.DDMFormField;
+import com.liferay.dynamic.data.mapping.model.DDMFormFieldValidation;
+import com.liferay.dynamic.data.mapping.model.DDMStructure;
+import com.liferay.dynamic.data.mapping.service.DDMStructureLocalService;
+import com.liferay.dynamic.data.mapping.test.util.DDMFormTestUtil;
+import com.liferay.dynamic.data.mapping.test.util.DDMStructureTestUtil;
+import com.liferay.dynamic.data.mapping.util.DDMFormSerializeUtil;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.change.tracking.CTModel;
+import com.liferay.portal.kernel.service.change.tracking.CTService;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.RandomTestUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.test.rule.PermissionCheckerMethodTestRule;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.UpgradeTestUtil;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Igor Costa
+ */
+@RunWith(Arquillian.class)
+public class DDMStructureEmptyValidationUpgradeProcessTest
+	extends BaseCTUpgradeProcessTestCase {
```

Comment:
Since we have multiple examples of DDMStructure and DDMStructureVersion, we don't want to duplicate code.

So, can you check this [PR](https://github.com/liferay-forms/liferay-portal/pull/2588).
Note:

- [v3_8_1/test/DDMStructureCTUpgradeProcessTest.java](https://github.com/liferay-forms/liferay-portal/pull/2588/files#diff-2bfd5ecd14944b9179c96dc2986bc051c4f94060d0ead12df31dbcd6846ddcc7)
- [v3_8_1/test/DDMStructureVersionDDMStructureCTUpgradeProcessTest.java](https://github.com/liferay-forms/liferay-portal/pull/2588/files#diff-3b4edc5788c5fec92f52ad89dabea9872af93200d6213ff6e361f8da4bf45cc2)

we should do similar to these examples.

Reply:
You should pay attention to [v3_8_1/test/util/DDMStructureCTUpgradeProcessTestUtil.java](https://github.com/liferay-forms/liferay-portal/pull/2588/files#diff-847db5b954d10a25b187011a9f2cd1466ccf4ef95b12a369bfc4efdf8e281d22) to avoid duplicate codes between the classes.

Reply:
Even though all classes test the same entities, each upgrade process has its own logic.
So you should pay attention to `getDDMForm()`. This method will contain the specific logic your upgrade is testing (similar to the `setDefinition()` method you created).

Reply:
You can cherry-pick that other PR to work on top of it :+1: